src/main/java/io/quarkus/search/app/fetching/FetchingService.java

```diff
@@ -103,7 +129,28 @@ private static boolean isZip(URI uri) {
                 && uri.getPath().endsWith(".zip");
     }
 
-    private <T> T gitClone(String name, URI gitUri, List<String> branches,
+    private GitCloneDirectory gitPull(GitCloneDirectory repository, Branches branches) {
+        Log.infof("Pulling changes for '%s'.", repository);
+        //String mainBranch = branches.get(0);
+        Log.infof("Pulling changes for current branch of '%s':'%s'.", repository, branches.sources());
+        try {
+            repository.git().pull()
+                    .setProgressMonitor(LoggerProgressMonitor.create(log, "Pulling '" + branches.sources() + "': "))
+                    .call();
+            repository.git().fetch().setForceUpdate(true)
+                    .setProgressMonitor(LoggerProgressMonitor.create(log, "Fetching '" + branches + "': "))
+                    .setRefSpecs(branches.asRefArray())
+                    .call();
+        } catch (RuntimeException | GitAPIException e) {
+            log.errorf(
+                    e, "Wasn't able to pull changes for branch '%s' in repository '%s': '%s", branches, repository,
+                    e.getMessage());
+        }
+        // we want "pagesTree" to get recreated to point to an updated ref:
+        return repository.refresh();
```

Comment:
This makes me think we probably don't want to cache the `Git`/`RevTree` and whatever they keep in memory... 

Maybe you could just cache the `CloseableDirectory`? Then upon cache hits you'd do a `Git.open()` and recreate the `GitCloneDirectory`. `QuarkusIO.close` would trigger `GitCloneDirectory.close()`, which would close the `Git` but would not close the `CloseableDirectory`, keeping it around for reuse.

BTW if you do this, you don't even need to represent the directory as a `CloseableDirectory` in `GitCloneDirectory`, a `Path` would be enough.